### PR TITLE
Unpark actual threads waiting on channels

### LINKED ISSUE #760: Potential inconsistency in Rust Oak Runtime thread handling
When calling `Runtime::wait_on_channel` from a Node

https://github.com/project-oak/oak/blob/70da1899692a7ceb9eb0875fb3ee001645adb04e/oak/server/rust/oak_runtime/src/runtime/mod.rs#L160

the "current thread" is looked up using `thread::current`

https://github.com/project-oak/oak/blob/70da1899692a7ceb9eb0875fb3ee001645adb04e/oak/server/rust/oak_runtime/src/runtime/mod.rs#L164

and eventually parked, if no message is readily available on the specified channel(s)

https://github.com/project-oak/oak/blob/70da1899692a7ceb9eb0875fb3ee001645adb04e/oak/server/rust/oak_runtime/src/runtime/mod.rs#L206

Then, when calling `Runtime::stop`, the thread corresponding to the Node instance is unparked

https://github.com/project-oak/oak/blob/70da1899692a7ceb9eb0875fb3ee001645adb04e/oak/server/rust/oak_runtime/src/runtime/mod.rs#L117-L120

I think the assumption here is that this would be the same single thread that is associated with the Node instance, which is the case for the current implementation.

The problem arises if (or rather, when) we introduce Nodes that may spawn additional threads, since it may happen that the call to `Runtime::wait_on_channel` does not come from the initial thread, and therefore unparking the initial thread would not actually have the desired effect.

I think the correct (future-proof) thing to do would be for `Runtime::wait_on_channel` to keep track of the actual waiting thread itself, perhaps in a separate field (perhaps of type `HashMap<ThreadId, JoinHandle>`), which should also be able to deal with the fact that multiple threads from the same node may be waiting at the same time.

@blaxill WDYT?

### PR DESCRIPTION
**Pull request by tiziano88**
Fixes #760

# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by tiziano88** [comment]
It would be nice to have at least a unit test to verify that this is actually solving the issue, but I cannot think of an easy way to write one.

Also there is some similarity with 

https://github.com/project-oak/oak/blob/70da1899692a7ceb9eb0875fb3ee001645adb04e/oak/server/rust/oak_runtime/src/runtime/channel.rs#L39

but they are not quite the same. Do you think we should be using `Weak<Thread>` for this PR too?

**Comment by tiziano88** [comment]
@blaxill actually changed things around now, PTAL, I am just reusing the existing map, and iterate over channels and then inside that over waiting threads, and unparking everything that is still upgradeable.

**Comment by blaxill** [comment]
This is a valid solution, another one would be for nodes to add all `JoinHandles` to `runtime::Node` and iterate over them when stopping.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 1]
```diff
@@ -107,20 +107,35 @@ impl Runtime {
...
+            {
+                if let Some(thread) = thread.upgrade() {
+                    thread.unpark();
+                }
+            }
+        }
+
+        let mut nodes = {
             let mut nodes = self.nodes.lock().unwrap();
             self.terminating.store(true, SeqCst);
```
Comment (blaxill) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs]: I think swapping the nodes out and setting terminating to true needs to occur before unparking the threads as the wait_on_channels loop will cause them to park again if runtime is not terminating.

Comment (blaxill) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs]: This is probably why CI is timing out.

Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs]: Good point, fixed, and added some comments, PTAL.

**Pull request merged**